### ISSUE TITLE
Support for Twitter API v2

### ISSUE THREAD

**Issue by Harmon758** [tweepy/tweepy#1472]
~~Now that [the new Twitter API has been launched](https://blog.twitter.com/developer/en_us/topics/tools/2020/introducing_new_twitter_api.html), support for it needs to be added in Tweepy.~~
~~This is planned to be part of v4.0.~~

Initial support for Twitter API v2 has been added with [Tweepy v4.0](https://github.com/tweepy/tweepy/releases/tag/v4.0.0).
Support for additional endpoints and features have been added with Tweepy [v4.1](https://github.com/tweepy/tweepy/releases/tag/v4.1.0), [v4.2](https://github.com/tweepy/tweepy/releases/tag/v4.2.0), [v4.3](https://github.com/tweepy/tweepy/releases/tag/v4.3.0), [v4.4](https://github.com/tweepy/tweepy/releases/tag/v4.4.0), and [v4.5](https://github.com/tweepy/tweepy/releases/tag/v4.5.0).

Support for new endpoints and streaming will be added with future v4.x releases.

~~v3.10 will be released first, as the last version to support Python 2.7 (and probably Python 3.5).
My goal is to get v3.10 out before the end of the year and then work on [dropping support for Python 2](https://github.com/tweepy/tweepy/issues/1253).~~
v3.10 has been released now, and support for Python 2.7 and 3.5 has been dropped.
~~Afterwards,~~ This will be the next priority, along with [support for asyncio](https://github.com/tweepy/tweepy/issues/732).

##### Filtered stream
- [X] [GET /2/tweets/search/stream](https://developer.twitter.com/en/docs/twitter-api/tweets/filtered-stream/api-reference/get-tweets-search-stream) (https://github.com/tweepy/tweepy/commit/86244c1a82a1852d04f3695b03201363f5d5eafd)
- [X] [GET /2/tweets/search/stream/rules](https://developer.twitter.com/en/docs/twitter-api/tweets/filtered-stream/api-reference/get-tweets-search-stream-rules) (https://github.com/tweepy/tweepy/commit/86244c1a82a1852d04f3695b03201363f5d5eafd)
- [X] [POST /2/tweets/search/stream/rules](https://developer.twitter.com/en/docs/twitter-api/tweets/filtered-stream/api-reference/post-tweets-search-stream-rules) (https://github.com/tweepy/tweepy/commit/86244c1a82a1852d04f3695b03201363f5d5eafd)

##### [Hide replies](https://twittercommunity.com/t/hide-replies-now-available-in-the-new-twitter-api/140996)
- [X] [PUT /2/tweets/:id/hidden](https://developer.twitter.com/en/docs/twitter-api/tweets/hide-replies/api-reference/put-tweets-id-hidden) (#1535)

##### Likes ([lookup](https://twittercommunity.com/t/announcing-twitter-api-v2-likes-lookup-and-blocks-lookup/154353) + [manage](https://twittercommunity.com/t/announcing-new-manage-likes-endpoints-for-thetwitter-api-v2/152780))
- [X] [DELETE /2/users/:id/likes/:tweet_id](https://developer.twitter.com/en/docs/twitter-api/tweets/likes/api-reference/delete-users-user_id-likes) (https://github.com/tweepy/tweepy/commit/5beaa24e4c39b9c6777c87dd7e0eeed9a9ee8355)
- [X] [GET /2/tweets/:id/liking_users](https://developer.twitter.com/en/docs/twitter-api/tweets/likes/api-reference/get-tweets-id-liking_users) (https://github.com/tweepy/tweepy/commit/a32e6cea08d170e618d46eabc82d673074555c3d)
- [X] [GET /2/users/:id/liked_tweets](https://developer.twitter.com/en/docs/twitter-api/tweets/likes/api-reference/get-users-id-liked_tweets) (https://github.com/tweepy/tweepy/commit/a32e6cea08d170e618d46eabc82d673074555c3d)
- [X] [POST /2/users/:id/likes](https://developer.twitter.com/en/docs/twitter-api/tweets/likes/api-reference/post-users-user_id-likes) (https://github.com/tweepy/tweepy/commit/5beaa24e4c39b9c6777c87dd7e0eeed9a9ee8355)

##### [Manage Tweets](https://twittercommunity.com/t/announcing-manage-tweets-endpoints-for-the-twitter-api-v2/161501)
- [X] [DELETE /2/tweets/:id](https://developer.twitter.com/en/docs/twitter-api/tweets/manage-tweets/api-reference/delete-tweets-id) (https://github.com/tweepy/tweepy/commit/7884e3a7253d9a821ff46160ec0d3811f299327f)
- [X] [POST /2/tweets](https://developer.twitter.com/en/docs/twitter-api/tweets/manage-tweets/api-reference/post-tweets) (https://github.com/tweepy/tweepy/commit/7884e3a7253d9a821ff46160ec0d3811f299327f)

##### [Retweets](https://twittercommunity.com/t/announcing-new-twitter-api-v2-retweets-endpoints/156827)
- [X] [DELETE /2/users/:id/retweets/:source_tweet_id](https://developer.twitter.com/en/docs/twitter-api/tweets/retweets/api-reference/delete-users-id-retweets-tweet_id) (https://github.com/tweepy/tweepy/commit/b4985e04acb87b791e4d0d3f45bd57e5410e7d48)
- [X] [GET /2/tweets/:id/retweeted_by](https://developer.twitter.com/en/docs/twitter-api/tweets/retweets/api-reference/get-tweets-id-retweeted_by) (https://github.com/tweepy/tweepy/commit/b4985e04acb87b791e4d0d3f45bd57e5410e7d48)
- [X] [POST /2/users/:id/retweets](https://developer.twitter.com/en/docs/twitter-api/tweets/retweets/api-reference/post-users-id-retweets) (https://github.com/tweepy/tweepy/commit/b4985e04acb87b791e4d0d3f45bd57e5410e7d48)

##### Search Tweets ([full-archive](https://twittercommunity.com/t/introducing-the-new-academic-research-product-track/148632) + recent)
- [X] [GET /2/tweets/search/all](https://developer.twitter.com/en/docs/twitter-api/tweets/full-archive-search/api-reference/get-tweets-search-all) (#1535)
- [X] [GET /2/tweets/search/recent](https://developer.twitter.com/en/docs/twitter-api/tweets/search/api-reference/get-tweets-search-recent) (#1535)

##### [Timelines](https://twittercommunity.com/t/user-tweet-timeline-and-user-mention-timeline-endpoints-available-for-early-access-in-the-twitter-api-v2/147115)
- [X] [GET /2/users/:id/mentions](https://developer.twitter.com/en/docs/twitter-api/tweets/timelines/api-reference/get-users-id-mentions) (#1535)
- [X] [GET /2/users/:id/tweets](https://developer.twitter.com/en/docs/twitter-api/tweets/timelines/api-reference/get-users-id-tweets) (#1535)
- ~~[GET /2/users/by/username/:username/mentions](https://developer.twitter.com/en/docs/twitter-api/tweets/timelines/api-reference/get-users-by-username-username-mentions)~~
- ~~[GET /2/users/by/username/:username/tweets](https://developer.twitter.com/en/docs/twitter-api/tweets/timelines/api-reference/get-users-by-username-username-tweets)~~

##### [Tweet counts](https://twittercommunity.com/t/introducing-new-tweet-counts-endpoints-to-the-twitter-api-v2/155997)
- [X] [GET /2/tweets/counts/all](https://developer.twitter.com/en/docs/twitter-api/tweets/counts/api-reference/get-tweets-counts-all) (https://github.com/tweepy/tweepy/commit/49156cb9f1d74d709480b8d746b01e755f5c312b)
- [X] [GET /2/tweets/counts/recent](https://developer.twitter.com/en/docs/twitter-api/tweets/counts/api-reference/get-tweets-counts-all) (https://github.com/tweepy/tweepy/commit/49156cb9f1d74d709480b8d746b01e755f5c312b)

##### Tweet lookup
- [X] [GET /2/tweets](https://developer.twitter.com/en/docs/twitter-api/tweets/lookup/api-reference/get-tweets) (#1535)
- [X] [GET /2/tweets/:id](https://developer.twitter.com/en/docs/twitter-api/tweets/lookup/api-reference/get-tweets-id) (#1535)

##### Volume streams
- [X] [GET /2/tweets/sample/stream](https://developer.twitter.com/en/docs/twitter-api/tweets/sampled-stream/api-reference/get-tweets-sample-stream) (https://github.com/tweepy/tweepy/commit/86244c1a82a1852d04f3695b03201363f5d5eafd)

##### Blocks ([lookup](https://twittercommunity.com/t/announcing-twitter-api-v2-likes-lookup-and-blocks-lookup/154353) + [manage](https://twittercommunity.com/t/announcing-manage-blocks-for-twitter-api-v2/152358))
- [X] [DELETE /2/users/:source_user_id/blocking/:target_user_id](https://developer.twitter.com/en/docs/twitter-api/users/blocks/api-reference/delete-users-user_id-blocking) (https://github.com/tweepy/tweepy/commit/18e0b1a37c36e270ccc9c50cb1343f03637f70e0)
- [X] [GET /2/users/:id/blocking](https://developer.twitter.com/en/docs/twitter-api/users/blocks/api-reference/get-users-blocking) (https://github.com/tweepy/tweepy/commit/a32e6cea08d170e618d46eabc82d673074555c3d)
- [X] [POST /2/users/:id/blocking](https://developer.twitter.com/en/docs/twitter-api/users/blocks/api-reference/post-users-user_id-blocking) (https://github.com/tweepy/tweepy/commit/18e0b1a37c36e270ccc9c50cb1343f03637f70e0)

##### Follows ([lookup](https://twittercommunity.com/t/introducing-the-v2-follow-lookup-endpoints/147118) + [manage](https://twittercommunity.com/t/introducing-the-new-manage-follows-endpoints-to-the-twitter-api-v2/149465))
- [X] [DELETE /2/users/:source_user_id/following/:target_user_id](https://developer.twitter.com/en/docs/twitter-api/users/follows/api-reference/delete-users-source_id-following) (#1535)
- [X] [GET /2/users/:id/followers](https://developer.twitter.com/en/docs/twitter-api/users/follows/api-reference/get-users-id-followers) (#1535)
- [X] [GET /2/users/:id/following](https://developer.twitter.com/en/docs/twitter-api/users/follows/api-reference/get-users-id-following) (#1535)
- ~~[GET /2/users/by/username/:username/followers](https://developer.twitter.com/en/docs/twitter-api/users/follows/api-reference/get-users-by-username-followers)~~
- ~~[GET /2/users/by/username/:username/following](https://developer.twitter.com/en/docs/twitter-api/users/follows/api-reference/get-users-by-username-following)~~
- [X] [POST /2/users/:id/following](https://developer.twitter.com/en/docs/twitter-api/users/follows/api-reference/post-users-source_user_id-following) (#1535)

##### Mutes ([lookup](https://twittercommunity.com/t/announcing-new-mutes-lookup-endpoint-for-the-twitter-api-v2/159680) + [manage](https://twittercommunity.com/t/announcing-new-manage-mutes-endpoints-for-the-twitter-api-v2/156261))
- [X] [DELETE /2/users/:source_user_id/muting/:target_user_id](https://developer.twitter.com/en/docs/twitter-api/users/mutes/api-reference/delete-users-user_id-muting) (https://github.com/tweepy/tweepy/commit/b20b236c4ca10a5376ea2f612a1ad324f21d7072)
- [X] [GET /2/users/:id/muting](https://developer.twitter.com/en/docs/twitter-api/users/mutes/api-reference/get-users-muting) (https://github.com/tweepy/tweepy/commit/00cdab8a86b98c6b02636fd511700fc73653e006)
- [X] [POST /2/users/:id/muting](https://developer.twitter.com/en/docs/twitter-api/users/mutes/api-reference/post-users-user_id-muting) (https://github.com/tweepy/tweepy/commit/b20b236c4ca10a5376ea2f612a1ad324f21d7072)

##### User lookup
- [X] [GET /2/users](https://developer.twitter.com/en/docs/twitter-api/users/lookup/api-reference/get-users) (#1535)
- [X] [GET /2/users/:id](https://developer.twitter.com/en/docs/twitter-api/users/lookup/api-reference/get-users-id) (#1535)
- [X] [GET /2/users/by](https://developer.twitter.com/en/docs/twitter-api/users/lookup/api-reference/get-users-by) (#1535)
- [X] [GET /2/users/by/username/:username](https://developer.twitter.com/en/docs/twitter-api/users/lookup/api-reference/get-users-by-username-username) (#1535)
- [X] [GET /2/users/me](https://developer.twitter.com/en/docs/twitter-api/users/lookup/api-reference/get-users-me) (https://github.com/tweepy/tweepy/commit/c49cbdfcbda48295591d731446cd03b2eb2332ae, https://github.com/tweepy/tweepy/commit/62b5b586e75a850427eabdf31448d73a9e564f66)

##### Search Spaces
- [X] [GET /2/spaces/search](https://developer.twitter.com/en/docs/twitter-api/spaces/search/api-reference/get-spaces-search) (https://github.com/tweepy/tweepy/commit/5c688922cf0f8e89d401ea5108b06aaa8c12d71b)

##### Spaces lookup ([buyers](https://twittercommunity.com/t/adding-new-functionality-to-spaces-endpoints/161959), [Tweets](https://twittercommunity.com/t/bringing-tweets-shared-in-a-space-and-rsvp-count-to-the-spaces-endpoints/166746))
- [X] [GET /2/spaces](https://developer.twitter.com/en/docs/twitter-api/spaces/lookup/api-reference/get-spaces) (https://github.com/tweepy/tweepy/commit/5c688922cf0f8e89d401ea5108b06aaa8c12d71b)
- [X] [GET /2/spaces/:id](https://developer.twitter.com/en/docs/twitter-api/spaces/lookup/api-reference/get-spaces-id) (https://github.com/tweepy/tweepy/commit/5c688922cf0f8e89d401ea5108b06aaa8c12d71b)
- [X] [GET /2/spaces/:id/buyers](https://developer.twitter.com/en/docs/twitter-api/spaces/lookup/api-reference/get-spaces-id-buyers) (https://github.com/tweepy/tweepy/commit/8bf58ca5754f72292d1d86367a02778af7a44f20)
- [X] [GET /2/spaces/:id/tweets](https://developer.twitter.com/en/docs/twitter-api/spaces/lookup/api-reference/get-spaces-id-tweets) (https://github.com/tweepy/tweepy/commit/c8d5d9ad2c2b3c6b58b987286726cad467ce888f)
- [X] [GET /2/spaces/by/creator_ids](https://developer.twitter.com/en/docs/twitter-api/spaces/lookup/api-reference/get-spaces-by-creator-ids) (https://github.com/tweepy/tweepy/commit/5c688922cf0f8e89d401ea5108b06aaa8c12d71b)

##### List Tweets lookup

- [X] [GET /2/lists/:id/tweets](https://developer.twitter.com/en/docs/twitter-api/lists/list-tweets/api-reference/get-lists-id-tweets) (https://github.com/tweepy/tweepy/commit/0aa2366a875464756507abf42709a3db676f4cee)

##### List follows

- [X] [DELETE /2/users/:id/followed_lists/:list_id](https://developer.twitter.com/en/docs/twitter-api/lists/list-follows/api-reference/delete-users-id-followed-lists-list_id) (https://github.com/tweepy/tweepy/commit/b1342bfc998bee334437f2b7a8d2aef4df7c3838)
- [X] [GET /2/lists/:id/followers](https://developer.twitter.com/en/docs/twitter-api/lists/list-follows/api-reference/get-lists-id-followers) (https://github.com/tweepy/tweepy/commit/0aa2366a875464756507abf42709a3db676f4cee)
- [X] [GET /2/users/:id/followed_lists](https://developer.twitter.com/en/docs/twitter-api/lists/list-follows/api-reference/get-users-id-followed_lists) (https://github.com/tweepy/tweepy/commit/0aa2366a875464756507abf42709a3db676f4cee)
- [X] [POST /2/users/:id/followed_lists](https://developer.twitter.com/en/docs/twitter-api/lists/list-follows/api-reference/post-users-id-followed-lists) (https://github.com/tweepy/tweepy/commit/b1342bfc998bee334437f2b7a8d2aef4df7c3838)

##### List lookup

- [X] [GET /2/lists/:id](https://developer.twitter.com/en/docs/twitter-api/lists/list-lookup/api-reference/get-lists-id) (https://github.com/tweepy/tweepy/commit/0aa2366a875464756507abf42709a3db676f4cee)
- [X] [GET /2/users/:id/owned_lists](https://developer.twitter.com/en/docs/twitter-api/lists/list-lookup/api-reference/get-users-id-owned_lists) (https://github.com/tweepy/tweepy/commit/0aa2366a875464756507abf42709a3db676f4cee)

##### List members

- [X] [DELETE /2/lists/:id/members/:user_id](https://developer.twitter.com/en/docs/twitter-api/lists/list-members/api-reference/delete-lists-id-members-user_id) (https://github.com/tweepy/tweepy/commit/b1342bfc998bee334437f2b7a8d2aef4df7c3838)
- [X] [GET /2/lists/:id/members](https://developer.twitter.com/en/docs/twitter-api/lists/list-members/api-reference/get-lists-id-members) (https://github.com/tweepy/tweepy/commit/0aa2366a875464756507abf42709a3db676f4cee)
- [X] [GET /2/users/:id/list_memberships](https://developer.twitter.com/en/docs/twitter-api/lists/list-members/api-reference/get-users-id-list_memberships) (https://github.com/tweepy/tweepy/commit/0aa2366a875464756507abf42709a3db676f4cee)
- [X] [POST /2/lists/:id/members](https://developer.twitter.com/en/docs/twitter-api/lists/list-members/api-reference/post-lists-id-members) (https://github.com/tweepy/tweepy/commit/b1342bfc998bee334437f2b7a8d2aef4df7c3838)

##### [Manage Lists](https://twittercommunity.com/t/announcing-new-twitter-api-v2-lists-endpoints/159980)

- [X] [DELETE /2/lists/:id](https://developer.twitter.com/en/docs/twitter-api/lists/manage-lists/api-reference/delete-lists-id) (https://github.com/tweepy/tweepy/commit/b1342bfc998bee334437f2b7a8d2aef4df7c3838)
- [X] [PUT /2/lists/:id](https://developer.twitter.com/en/docs/twitter-api/lists/manage-lists/api-reference/put-lists-id) (https://github.com/tweepy/tweepy/commit/b1342bfc998bee334437f2b7a8d2aef4df7c3838)
- [X] [POST /2/lists](https://developer.twitter.com/en/docs/twitter-api/lists/manage-lists/api-reference/post-lists) (https://github.com/tweepy/tweepy/commit/b1342bfc998bee334437f2b7a8d2aef4df7c3838)

##### Pinned Lists

- [X] [DELETE /2/users/:id/pinned_lists/:list_id](https://developer.twitter.com/en/docs/twitter-api/lists/pinned-lists/api-reference/delete-users-id-pinned-lists-list_id) (https://github.com/tweepy/tweepy/commit/b1342bfc998bee334437f2b7a8d2aef4df7c3838)
- [X] [GET /2/users/:id/pinned_lists](https://developer.twitter.com/en/docs/twitter-api/lists/pinned-lists/api-reference/get-users-id-pinned_lists) (https://github.com/tweepy/tweepy/commit/0aa2366a875464756507abf42709a3db676f4cee)
- [X] [POST /2/users/:id/pinned_lists](https://developer.twitter.com/en/docs/twitter-api/lists/pinned-lists/api-reference/post-users-id-pinned-lists) (https://github.com/tweepy/tweepy/commit/b1342bfc998bee334437f2b7a8d2aef4df7c3838)

##### [Batch Compliance](https://twittercommunity.com/t/new-batch-compliance-endpoints-available-on-the-twitter-api-v2/158534)
- [X] [GET /2/compliance/jobs](https://developer.twitter.com/en/docs/twitter-api/compliance/batch-compliance/api-reference/get-compliance-jobs) (https://github.com/tweepy/tweepy/commit/6ca75e1b69a1085ac43215b30fc8269b91faa790)
- [X] [GET /2/compliance/jobs/:id](https://developer.twitter.com/en/docs/twitter-api/compliance/batch-compliance/api-reference/get-compliance-jobs-id) (https://github.com/tweepy/tweepy/commit/6ca75e1b69a1085ac43215b30fc8269b91faa790)
- [X] [POST /2/compliance/jobs](https://developer.twitter.com/en/docs/twitter-api/compliance/batch-compliance/api-reference/post-compliance-jobs) (https://github.com/tweepy/tweepy/commit/6ca75e1b69a1085ac43215b30fc8269b91faa790)

#### Addendum

- [X] [Media `alt_text` field](https://twittercommunity.com/t/media-alt-text-field-now-available-in-twitter-api-v2/157939) (https://github.com/tweepy/tweepy/commit/379e99a672537776ac0e160999967b5efce29305)
- [X] [`state` parameter optional for spaces search endpoint](https://twitter.com/i_am_daniele/status/1443210956541100033) (https://github.com/tweepy/tweepy/commit/e61d5d6f017db7f28871331b515fcfd87666f352)
- [X] [Space `ended_at` and `topic_ids` fields](https://twittercommunity.com/t/adding-new-functionality-to-spaces-endpoints/161959) (https://github.com/tweepy/tweepy/commit/c89a233a96142ce01ecd5e1372c3fdce45c6601f)
- [X] [OAuth 2.0 Authorization Code Flow with Proof Key for Code Exchange (PKCE)](https://twittercommunity.com/t/announcing-oauth-2-0-general-availability/163555) (https://github.com/tweepy/tweepy/commit/2b835073cb193ca6f2849c8cb6ef4322e5b16f24, https://github.com/tweepy/tweepy/commit/8f384294405c6d14507441a5d1a7040d927b3fc2)
- [X] [Updates to Retweets lookup and Likes lookup endpoints](https://twittercommunity.com/t/updates-to-retweets-lookup-and-likes-lookup-endpoints/165327) (https://github.com/tweepy/tweepy/commit/3479e56a0fff02594ab6e77ca8227e65b51b46f7, https://github.com/tweepy/tweepy/commit/bdd6b55d7cb075fc2e4c7cb56a061c552ca106fe)
- [X] [`sort_order` parameter for search endpoints](https://twittercommunity.com/t/introducing-the-sort-order-parameter-for-search-endpoints-in-the-twitter-api-v2/166377) (https://github.com/tweepy/tweepy/commit/bd202e5be19670116ebe98329d388b4821d5f3eb, https://github.com/tweepy/tweepy/commit/8b47170e7c20127add3e67190890d2a4cef92266)
- [X] [Space `subscriber_count` field](https://twittercommunity.com/t/bringing-tweets-shared-in-a-space-and-rsvp-count-to-the-spaces-endpoints/166746) (https://github.com/tweepy/tweepy/commit/1ffc8cda6d2b9670b86610859f21ac05d726eaf9)

###### COVID-19 stream (#1790)

- [GET /labs/1/tweets/stream/compliance](https://developer.twitter.com/en/docs/twitter-api/tweets/covid-19-stream/api-reference/get-tweets-stream-compliance)
- [GET /labs/1/tweets/stream/covid19](https://developer.twitter.com/en/docs/twitter-api/tweets/covid-19-stream/api-reference/get-tweets-stream-covid19)

<details>
  <summary>Additional Resources</summary>

  https://developer.twitter.com/en/docs/twitter-api/early-access
  https://developer.twitter.com/en/docs/twitter-api/migrate
  https://developer.twitter.com/en/products/twitter-api/early-access/guide
  https://blog.twitter.com/developer/en_us/topics/tools/2020/introducing_new_twitter_api.html
  https://twittercommunity.com/t/announcing-early-access-to-the-next-generation-of-the-twitter-api/139612

  https://twittercommunity.com/t/blocks-and-likes-endpoints-are-now-complete-on-the-twitter-api-v2/154434

  https://blog.twitter.com/developer/en_us/topics/tools/2021/shape-the-future-of-twitter-spaces

  https://twittercommunity.com/t/announcing-oauth-2-0-beta/159189
</details>

<details>
  <summary>Old Postscript</summary>

  I'd like to apologize for the delay in responding to @andypiper and others about this in #1345, #1396, and #1423, as well as on Discord. I've (I think understandably) fallen behind a bit this year in my plans for Tweepy and in keeping up with issues and Discord messages, but I'm working on catching back up.
</details>

**Comment by Harmon758** [comment]
I've created an initial implementation in an api-v2 branch and PR #1535.

**Comment by azharmunir43** [comment]
hey @Harmon758 thanks for this, can we get some idea about when is it going to be released? Thanks

**Comment by Harmon758** [comment]
Most of the endpoints should now be part of the api-v2 branch / PR #1535.
Once that PR is ready, I'll take it out of its draft state and leave it open for a few weeks for people to test and review.
Afterwards, it'll be merged into the master branch.
As for v4.0 itself, there's other features (https://github.com/tweepy/tweepy/milestone/11) that I want to include in it before it's released.

**Comment by Harmon758** [comment]
The api-v2 branch / pull request #1535 should be complete now.
Any feedback or review would be appreciated.

**Comment by laurarosenberg** [comment]
Hey, 
I am not a programming professional and now I have added the new api-v2-branch from tweepy. Somehow I don't understand how I can send queries with search_full_archive now? I have adapted my script:

```
try: 
    auth = tweepy.AppAuthHandler(Credentials.consumer_key2, Credentials.consumer_secret2) 
    api = tweepy.API(auth,wait_on_rate_limit=True)
except Exception as e: # catch connection errors 
        print("Authentication connection error: {}".format(e.reason))    

try: 
    tweets = tweepy.Cursor(api.search_full_archive, 
                label = ''
                query='\"{}\"'.format(searchWords), 
                lang="en", 
                fromDate= data_since_pro
                toDate='202005030000'
                ).items(numTweets)
except Exception as e: # catch connection error 
    print("API connection error: {}".format(e.reason))  

# store these tweets in a Python list
tweet_list = list(tweets)
  
# retrieve information from tweets iterable object
for tweet in tweet_list:
    print(tweet.text)
```

What am I doing wrong? 
How can I ask for concrete objects afterwards? Like count_likes, or retweets?

Translated with www.DeepL.com/Translator (free version)

**Comment by Harmon758** [comment]
The API v2 interface is the `Client` class and the pagination interface for it is `Paginator`. See #1535.
[`API.search_full_archive`](https://tweepy.readthedocs.io/en/latest/api.html#tweepy.API.search_full_archive) uses the [premium search endpoints](https://developer.twitter.com/en/docs/twitter-api/premium/search-api/api-reference/premium-search), part of API v1.1, not v2.
This is unrelated to this issue.

You need to specify your environment label. `lang` is not a valid parameter for the method/endpoint, and you should not assume that any exception raised is a connection error.

**Comment by Harmon758** [comment]
The api-v2 branch / pull request #1535 has been merged into the master branch.

**Comment by Harmon758** [comment]
`Client.like` and `Client.unlike` have been added with https://github.com/tweepy/tweepy/commit/5beaa24e4c39b9c6777c87dd7e0eeed9a9ee8355.

**Comment by Harmon758** [comment]
`Client.block` and `Client.unblock` have been added with https://github.com/tweepy/tweepy/commit/18e0b1a37c36e270ccc9c50cb1343f03637f70e0.

**Comment by Bowrna** [comment]
hi @Harmon758 how to pick up one of these tasks and start contributing? I am new to contributing to open source projects. Could you point me to link or doc that can help me to get started?

**Comment by Harmon758** [comment]
Pull requests are welcome. You can find many resources for how to make pull requests and the best practices for that with a simple web search.

Otherwise, for Tweepy, there's not much additional information you need to know besides how to run tests, which is [covered in the documentation](https://tweepy.readthedocs.io/en/latest/running_tests.html).

As for Twitter API v2 support specifically, I plan on adding most of the new endpoints myself this week.

You should [start a new discussion](https://github.com/tweepy/tweepy/discussions/new) if you have further questions, as this isn't specifically relevant to this issue.

**Comment by Harmon758** [comment]
`Client.get_liking_users`, `Client.get_liked_tweets`, and `Client.get_blocked` have been added with https://github.com/tweepy/tweepy/commit/a32e6cea08d170e618d46eabc82d673074555c3d.

**Comment by Harmon758** [comment]
`Client.get_all_tweets_count` and `Client.get_recent_tweets_count` have been added with https://github.com/tweepy/tweepy/commit/49156cb9f1d74d709480b8d746b01e755f5c312b.

**Comment by Harmon758** [comment]
`Client.unmute` and `Client.mute` have been added with https://github.com/tweepy/tweepy/commit/b20b236c4ca10a5376ea2f612a1ad324f21d7072.

**Comment by Harmon758** [comment]
`Client.unretweet`, `Client.get_retweeters`, and `Client.retweet` have been added with https://github.com/tweepy/tweepy/commit/b4985e04acb87b791e4d0d3f45bd57e5410e7d48.

**Comment by Harmon758** [comment]
The `alt_text` field for Media has been added with https://github.com/tweepy/tweepy/commit/379e99a672537776ac0e160999967b5efce29305.

**Comment by Harmon758** [comment]
Newer endpoints and streaming will be added as part of later v4.x releases.
I want to get v4.0 out with initial support for Twitter API v2 and the tons of other changes it includes.

**Comment by Harmon758** [comment]
[v4.0.0](https://github.com/tweepy/tweepy/releases/tag/v4.0.0) (https://github.com/tweepy/tweepy/discussions/1666) has now been released with support for Twitter API v2 🎉

**Comment by Harmon758** [comment]
Support for Spaces, with a `Space` model, `Client.search_spaces`, `Client.get_spaces`, and `Client.get_space`, has been added with https://github.com/tweepy/tweepy/commit/5c688922cf0f8e89d401ea5108b06aaa8c12d71b.

**Comment by Harmon758** [comment]
[v4.1.0](https://github.com/tweepy/tweepy/releases/tag/v4.1.0) (#1684) has now been released with support for Spaces, batch compliance, and mutes lookup (as well as Python 3.10) 🎉

**Comment by Harmon758** [comment]
[v4.2.0](https://github.com/tweepy/tweepy/releases/tag/v4.2.0) (#1694) has now been released with support for managing lists and the `state` parameter change for Spaces search 🎉

**Comment by Harmon758** [comment]
`Client.delete_tweet` and `Client.create_tweet` have been added with https://github.com/tweepy/tweepy/commit/7884e3a7253d9a821ff46160ec0d3811f299327f.

**Comment by Harmon758** [comment]
[v4.3.0](https://github.com/tweepy/tweepy/releases/tag/v4.3.0) (#1701) has been released with support for the Twitter API v2 manage Tweets endpoints launched today 🎉

**Comment by goyal-aman** [comment]
@Harmon758 I am interested in contributing. Though this would be my first time contributing to opensource. Any suggestions where to start?

**Comment by Harmon758** [comment]
Pull requests are welcome. You can find many resources for how to make pull requests and the best practices for that with a simple web search.

Otherwise, for Tweepy, there's not much additional information you need to know besides how to run tests, which is [covered in the documentation](https://tweepy.readthedocs.io/en/latest/running_tests.html).

As for Twitter API v2 support specifically, I'm working on adding support for the the last streaming endpoints.

You should [start a new discussion](https://github.com/tweepy/tweepy/discussions/new) if you have further questions, as this isn't specifically relevant to this issue.

**Comment by Harmon758** [comment]
A `List` model, `Client.get_list_tweets`, `Client.get_list_followers`, `Client.get_followed_lists`, `Client.get_list`, `Client.get_owned_lists`, `Client.get_list_members`, `Client.get_list_memberships`, and `Client.get_pinned_lists` have been added with https://github.com/tweepy/tweepy/commit/0aa2366a875464756507abf42709a3db676f4cee.

**Comment by Harmon758** [comment]
`Client.get_space_buyers` has been added with https://github.com/tweepy/tweepy/commit/8bf58ca5754f72292d1d86367a02778af7a44f20.

**Comment by Harmon758** [comment]
`Space.ended_at` and `Space.topic_ids` have been added with https://github.com/tweepy/tweepy/commit/c89a233a96142ce01ecd5e1372c3fdce45c6601f.

**Comment by Harmon758** [comment]
[v4.4.0](https://github.com/tweepy/tweepy/releases/tag/v4.4.0) (#1705) has been released with support for the new List lookup and Space endpoints, as well as the new Space object fields.

**Comment by Harmon758** [comment]
`Oauth2Handler` has been added with https://github.com/tweepy/tweepy/commit/2b835073cb193ca6f2849c8cb6ef4322e5b16f24, and `Client.get_me` has been added with https://github.com/tweepy/tweepy/commit/c49cbdfcbda48295591d731446cd03b2eb2332ae and https://github.com/tweepy/tweepy/commit/62b5b586e75a850427eabdf31448d73a9e564f66.

**Comment by Harmon758** [comment]
[v4.5.0](https://github.com/tweepy/tweepy/releases/tag/v4.5.0) (#1786) has been released with support for OAuth 2.0 Authorization Code Flow with PKCE and the new authenticated user lookup endpoint 🎉

**Comment by Harmon758** [comment]
Support for the new `sort_order` parameter for `Client.search_all_tweets` and `Client.search_recent_tweets` has been added with https://github.com/tweepy/tweepy/commit/bd202e5be19670116ebe98329d388b4821d5f3eb and https://github.com/tweepy/tweepy/commit/8b47170e7c20127add3e67190890d2a4cef92266, respectively.

**Comment by Harmon758** [comment]
`Client.get_space_tweets` has been added with https://github.com/tweepy/tweepy/commit/c8d5d9ad2c2b3c6b58b987286726cad467ce888f.

**Comment by Harmon758** [comment]
Support for streaming, with `StreamingClient`, `StreamResponse`, and `StreamRule`, has been added with https://github.com/tweepy/tweepy/commit/86244c1a82a1852d04f3695b03201363f5d5eafd.

**Comment by Harmon758** [comment]
`Space.subscriber_count` has been added with https://github.com/tweepy/tweepy/commit/1ffc8cda6d2b9670b86610859f21ac05d726eaf9.

**Comment by Harmon758** [comment]
[v4.6.0](https://github.com/tweepy/tweepy/releases/tag/v4.6.0) (#1824) has been released with support for streaming, the new Retweets lookup and Likes lookup endpoint parameters, the new `sort_order` parameter for search endpoints, and the new Tweets shared in a Space endpoint and RSVP count field for Spaces 🎉

With this release, Tweepy now fully supports all of Twitter API v2, almost exactly a year after the initial implementation for Twitter API v2 support was pushed. New individual issues will be created for any future new Twitter API v2 endpoints and features.

(The COVID-19 stream is the only remaining active portion of Labs and support for it is being tracked in #1790.)

**Comment by dhjw** [comment]
Need media by the end of the month #2217. Any interest?
